modules/EvseManager/IECStateMachine.cpp

```diff
@@ -0,0 +1,451 @@
+    {
+        std::scoped_lock lock(state_mutex);
+        // Only set the flags here in case of power on.
+        power_on_allowed = value;
+        power_on_reason = reason;
+        // In case of power off, we can directly forward this to the BSP driver here
+        if (not power_on_allowed) {
+            call_allow_power_on_bsp(false);
+        }
+    }
+    // The actual power on will be handled in the state machine to verify it is in the correct CP state etc.
+    feed_state_machine();
+}
+
+// Private member function used to actually call the BSP driver's allow_power_on
+// No need to lock mutex as this will be called from state machine or locked context only
+void IECStateMachine::call_allow_power_on_bsp(bool value) {
+    if (not value) {
+        power_on_allowed = false;
+        power_on_reason = types::evse_board_support::Reason::PowerOff;
+    }
+    r_bsp->call_allow_power_on({value, power_on_reason});
+}
+
+// High level state machine requests reading PP ampacity value.
+// We forward this request to the BSP driver. The high level state machine will never call this if it is not used
+// (e.g. in DC or AC tethered charging)
+double IECStateMachine::read_pp_ampacity() {
+    auto a = r_bsp->call_ac_read_pp_ampacity();
+    switch (a.ampacity) {
+    case types::board_support_common::Ampacity::A_13:
+        return 13.;
+    case types::board_support_common::Ampacity::A_20:
+        return 20.;
+    case types::board_support_common::Ampacity::A_32:
+        return 32.;
+    case types::board_support_common::Ampacity::A_63_3ph_70_1ph:
+        if (three_phases) {
+            return 63.;
+        } else {
+            return 70.;
+        }
+    default:
+        return 0.;
+    }
+}
+
+// Forward special replug request. Only for testing setups.
+void IECStateMachine::evse_replug(int ms) {
+    r_bsp->call_evse_replug(ms);
+}
+
+// Forward special request to switch the number of phases during charging. BSP will need to implement a special
+// sequence to not destroy cars.
+void IECStateMachine::switch_three_phases_while_charging(bool n) {
+    r_bsp->call_ac_switch_three_phases_while_charging(n);
+}
+
+// Forwards config parameters from EvseManager module config to BSP
+void IECStateMachine::setup(bool three_phases, bool has_ventilation, std::string country_code) {
+    r_bsp->call_setup(three_phases, has_ventilation, country_code);
+    this->has_ventilation = has_ventilation;
+    this->three_phases = three_phases;
+}
+
+// enable/disable the charging port and CP signal
+void IECStateMachine::enable(bool en) {
+    r_bsp->call_enable(en);
+}
+
+// Forward the over current detection limit to the BSP. Many BSP MCUs monitor the charge current and trigger a fault
+// in case of over current. This sets the target charging current value to be used in OC detection. It cannot be
+// derived from the PWM duty cycle, use this value instead.
+void IECStateMachine::set_overcurrent_limit(double amps) {
```

Comment:
is there a reason amps is not protected by the mutex?

Reply:
Looking at the current code it is ok as far as I can see, last_amps is an atomic double